src/core/Grappa.sol

```diff
@@ -413,62 +460,236 @@ contract Grappa is OwnableUpgradeable, ReentrancyGuardUpgradeable, UUPSUpgradeab
 
         if (block.timestamp < expiry) revert GP_NotExpired();
 
-        address engine = engines[_tokenId.parseEngineId()];
+        (address oracle, address engine, address underlying,, address strike,, address collateral, uint8 collateralDecimals) =
+            getDetailFromProductId(productId);
+
+        // expiry price of underlying, denominated in strike (usually USD), with {UNIT_DECIMALS} decimals
+        uint256 expiryPrice = _getCashSettlementPrice(oracle, underlying, strike, expiry);
+
+        // cash value denominated in strike (usually USD), with {UNIT_DECIMALS} decimals
+        uint256 cashValue;
+        if (tokenType == TokenType.CALL) {
+            cashValue = MoneynessLib.getCallCashValue(expiryPrice, strikePrice);
+        } else if (tokenType == TokenType.PUT) {
+            cashValue = MoneynessLib.getPutCashValue(expiryPrice, strikePrice);
+        } else if (tokenType == TokenType.CALL_SPREAD) {
+            cashValue = MoneynessLib.getCashValueDebitCallSpread(expiryPrice, strikePrice, reserved);
+        } else if (tokenType == TokenType.PUT_SPREAD) {
+            cashValue = MoneynessLib.getCashValueDebitPutSpread(expiryPrice, strikePrice, reserved);
+        }
 
-        if (settlementType == SettlementType.CASH) {
-            settlement.payoutPerToken = IMarginEngine(engine).getCashSettlementPerToken(_tokenId);
+        // the following logic convert cash value (amount worth) if collateral is not strike:
+        if (collateral == underlying) {
+            // collateral is underlying. payout should be devided by underlying price
+            cashValue = cashValue.mulDivDown(UNIT, expiryPrice);
+        } else if (collateral != strike) {
+            // collateral is not underlying nor strike
+            uint256 collateralPrice = _getCashSettlementPrice(oracle, collateral, strike, expiry);
+            cashValue = cashValue.mulDivDown(UNIT, collateralPrice);
+        }
 
-            if (settlement.payoutPerToken != 0) settlement.payoutAssetId = _tokenId.parseCollateralId();
-        } else if (settlementType == SettlementType.PHYSICAL) {
-            settlement = IPhysicalSettlement(engine).getPhysicalSettlementPerToken(_tokenId);
+        payoutPerToken = cashValue.convertDecimals(UNIT_DECIMALS, collateralDecimals);
+
+        return (engine, collateral, payoutPerToken);
+    }
+
+    /**
+     * @dev calculate the debt and payout for one physically settled option token
+     * @param _tokenId  token id of option token
+     * @return settlement struct
+     */
+    function _getPhysicalSettlementPerToken(uint256 _tokenId) internal view virtual returns (Settlement memory settlement) {
+        (TokenType tokenType,, uint40 productId, uint64 expiry, uint64 strikePrice,) = _tokenId.parseTokenId();
+
+        if (block.timestamp < expiry) revert GP_NotExpired();
+
+        (, uint8 engineId, uint8 underlyingId, uint8 strikeId,) = ProductIdUtil.parseProductId(productId);
```

Comment:
Same as above: I think this is implying that all engine with physical settlement has to use socialized payout. It should be abstract away from Grappa. (it has nothing to do with the payout, but only the "potential impact" of the payout)